plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt

```diff
@@ -45,23 +49,29 @@ interface FieldEmbedding {
             AccessPolicy.ALWAYS_READABLE, AccessPolicy.ALWAYS_WRITEABLE -> null
         }
 
-    fun accessInvariantsForPredicate(): TypeInvariantEmbedding? =
-        when (accessPolicy) {
-            AccessPolicy.ALWAYS_READABLE -> FieldAccessTypeInvariantEmbedding(this, PermExp.WildcardPerm())
-            AccessPolicy.ALWAYS_INHALE_EXHALE, AccessPolicy.ALWAYS_WRITEABLE -> null
-        }
+    fun accessInvariantsForPredicate(): TypeInvariantEmbedding? = null
 }
 
 class UserFieldEmbedding(
     override val name: ScopedKotlinName,
     override val type: TypeEmbedding,
     override val symbol: FirPropertySymbol,
+    override val isUnique: Boolean,
 ) : FieldEmbedding {
     override val viperType = Type.Ref
     override val accessPolicy: AccessPolicy = if (symbol.isVal) AccessPolicy.ALWAYS_READABLE else AccessPolicy.ALWAYS_INHALE_EXHALE
+    override val includeInUniquePredicate: Boolean
+        get() = isUnique || type.isPrimitive
     override val unfoldToAccess: Boolean
         get() = accessPolicy == AccessPolicy.ALWAYS_READABLE
     override val includeInShortDump: Boolean = true
+
+    override fun accessInvariantsForPredicate(): TypeInvariantEmbedding =
+        if (symbol.isVal) {
+            FieldAccessTypeInvariantEmbedding(this, PermExp.WildcardPerm())
+        } else {
+            FieldAccessTypeInvariantEmbedding(this, PermExp.FullPerm())
+        }
```

Comment:
This refers to the permissions inhaled, right?

Reply:
This tells us the way the field is accessed within the body of a predicate.

Reply:
I'm sorry, I don't follow.  Could you explain where it shows up?